Firestore/core/src/firebase/firestore/util/comparison.h

```diff
@@ -271,6 +272,27 @@ struct Comparator {
   }
 };
 
+/**
+ * A Comparator whose behavior is defined by a std::function.
+ */
+template <typename T>
+class FunctionComparator {
```

Comment:
C++ template code essentially is duck typing. Any code that operates on a comparator can't express any type constraint on its template parameter, and there need not be any relationship between the types that might be substituted in various instantiations of the template.

For more background, see ye olde article on the topic: http://www.drdobbs.com/templates-and-duck-typing/184401971

Reply:
Hrm...  I see.  Thanks.  Part of my confusion is that we _are_ using inheritance between DocumentComparator and FunctionComparator, but not between FunctionComparator and Comparator...  but I guess that's just for the sake of sharing code rather than a desire to implement a type-based contract.

Anyway, carry on!

Reply:
I can't say I can really articulate why I want `DocumentComparator` to be its own type rather than just an alias for `FunctionComparator<FSTDocument*>`. The way I did it was the most expeditious way to make that new type. An alternative would be to define it with a `using` statement like this:

```c++
using DocumentComparator = FunctionComparator<FSTDocument*>;
```

Would that help make things less confusing?

Reply:
Hrm.  Then where would `DocumentComparator::ByKey()` live?

In any case, I don't think it's worth worrying about.  Feel free to keep it as-is.  I think this is just part of my C++ acclimation process.

Reply:
Oh yeah, that's exactly why I wanted it to be its own type :-).